File: parameter_inference/gillespie-sampler.cpp
#include <cstdio>
#include <cstdlib>
#include <iostream>
#include <vector>

#include <gsl/gsl_rng.h>
#include <gsl/gsl_randist.h>

#include <thread>
#include <algorithm>

#include "../gillespie-algorithm.hpp"

// A Gillespie algorithm simulation of tumour suppressor loss
// This will generate data for the maximum likelihood section

std::map<int,std::vector<double>> generate_dataset(int seed, int runs) {
    // System coefficients:
    double rloh = 5.0e-7;
    double mu = 5.0e-8;

    Model model(5);
    model.m_migr[0][1] = mu;
    model.m_migr[0][2] = rloh;
    model.m_migr[1][3] = 0.5 * mu;
    model.m_migr[1][4] = 0.5 * rloh;
    model.m_migr[2][4] = 0.5 * mu;
    // birth and death rates:
    model.m_birth = {0, 0.05, 0.03, 0, 0};
    model.m_death = {0, 0, 0, 0, 0};
    model.m_initial_pops = {1e6, 0, 0, 0, 0};

    std::vector<int> final_vertices = {3, 4};

    // run some simulations and store the time and final node in
    // all_times:
    std::vector<std::pair<double,int>> all_times;
    times_to_final_vertices(model, seed, runs, final_vertices, all_times);

    std::map<int,std::vector<double>> all_times_flipped;
    for (auto& entry : all_times) {
        all_times_flipped[entry.second].push_back(entry.first);
    }

    return all_times_flipped;
}

int main(int argc, char* argv[]) {
    if (argc < 3) {
        std::cout << "Call this program with: " << std::endl;
        std::cout << "    ./gillespie-sampler seed runs";
        std::cout << std::endl;
    }

    int seed = atoi(argv[1]);
    int runs = atoi(argv[2]);

    std::map<int,std::vector<double>> all_times = generate_dataset(seed, runs);

    std::cout << "age, node," << std::endl;
    double age_max = 0;
    for (auto& pair : all_times) {
        for (auto& age : pair.second) {
            std::cout << age << ", " << pair.first << "," << std::endl;
            age_max += (age > age_max) * (age - age_max);
        }
        std::sort((pair.second).begin(), (pair.second).end());
    }
    std::cout << std::endl;

    std::cout << "\nSurvival curves:" << std::endl;
    size_t reference_pop = all_times[3].size() + all_times[4].size();

    for (auto& pair : all_times) {
        std::cout << "Type: " << pair.first << std::endl;
        print_kaplan_meier(age_max, pair.second, reference_pop);
    }

    return 0;
}

File: src/python/Test/gillespie_sampler.py
import pyffwd
import sys

def main():
    if len(sys.argv) < 3:
        print("Call this program with:")
        print("    python gillespie_sampler.py seed runs")
        return 1
    
    seed = int(sys.argv[1])
    runs = int(sys.argv[2])

    # System coefficients:
    rloh = 5.26337e-7
    mu = 2.16427e-8

    model = pyffwd.Model(5)
    model.m_migr = [
        {1: mu, 2: rloh}, # From vertex 0 to 1 and 2
        {3: 0.5 * mu, 4: 0.5 * rloh}, # From vertex 1 to 3 and 4
        {4: 0.5 * mu} # From vertex 2 to 4
    ]

    # Convert lists to RealVector
    model.m_birth = pyffwd.list_to_vector([0, 0.05, 0.03, 0, 0])
    model.m_death = pyffwd.list_to_vector([0, 0, 0, 0, 0])
    model.m_initial_pops = pyffwd.list_to_vector([1e6, 0, 0, 0, 0])

    final_vertices = [3, 4]

    # Run some simulations and store the time and final node in all_times:
    r = pyffwd.seed_gsl_rng(seed)
    all_times = []

    sim_state = pyffwd.GillespieInstance(model) # causes segfault FIXME
    print("hello")
    all_times = pyffwd.first_passage_time_multiple(r, model, final_vertices)

    print("wololo")

    pyffwd.times_to_final_vertices(model, seed, runs, final_vertices, all_times)

    print("age, node,")
    for pair in all_times:
        print(f"{pair[0]}, {pair[1]},")

    return 0

if __name__ == "__main__":
    main()
